### PR TITLE
[TokenClassification] Label realignment for subword aggregation

### LINKED ISSUE #10263: NER label re-alignment always expects B labelled first sub-words
## Environment info

- `transformers` version: 4.3.1
- Platform: Darwin-19.6.0-x86_64-i386-64bit
- Python version: 3.7.7
- PyTorch version (GPU?): 1.7.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help

- bert, tokenizers, pipelines: @LysandreJik
- trainer, maintained examples: @sgugger

## Information

Model I am using (Bert, XLNet ...): [DistilBERT fine-tuned for conll03](https://huggingface.co/elastic/distilbert-base-cased-finetuned-conll03-english)

The problem arises when using:
* [x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

1. Fine-tune a BERT model for NER/conll03 using the `run_ner.py` example script, all default values
2. Correct the label alignments, see [config.json](https://huggingface.co/elastic/distilbert-base-cased-finetuned-conll03-english/blob/main/config.json)
3. Infer using entities that have not been seen at training time, and are composed of multiple word-parts as defined by WordPiece (my assumption as to the cause).
4. Sub-words are labelled but pipeline re-grouping/label alignment relies on perfect sub-word labelling:

E.g. Accenture → A ##cc ##ent ##ure → B-ORG O O O → A (ORG)
E.g. Max Mustermann → Max Must ##erman ##n → B-PER I-PER I-PER O → Max Musterman (PER)
E.g. Elasticsearch → El ##astic ##sea #rch → O O I-MISC O → ##sea (MISC)

## Expected behavior

I would expect that the realignment takes the label from the first word part or the best scoring sub-word part and propogates that label to the entire word, never returning sub-words. The default in `run_ner.py` is to use a padded sub-word label at training as per the BERT paper, but I've not tried setting that to `False` yet as that's not the typical/standard practice.

E.g. Accenture → A ##cc ##ent ##ure → B-ORG O O O → Accenture (ORG)
E.g. Max Mustermann → Max Must ##erman ##n → B-PER I-PER I-PER O → Max Mustermann (PER)
E.g. Elasticsearch → El ##astic ##sea #rch → O O I-MISC O → Elasticsearch (MISC)

I'll add that it seems odd that this business logic is in the `pipeline`. When evaluating on conll03, I assume we are using the sub-words/first word, but this realignment should be considered during evaluation. As-is, I suspect the recall is lower than it should be.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Tentative to replace #11622 

- Added `AggregationStrategy`
- `ignore_subwords` and `grouped_entities` arguments are now fused
  into `aggregation_strategy`. It makes more sense anyway because
  `ignore_subwords=True` with `grouped_entities=False` did not have a
  meaning anyway.
- Added 2 new ways to aggregate which are MAX, and AVERAGE
- AVERAGE requires a bit more information than the others, for now this
case is slightly specific, we should keep that in mind for future
changes.
- Testing has been modified to reflect new argument, and to check the
correct deprecation and the new aggregation_strategy.
- Put the testing argument and testing results for aggregation_strategy,
close together, so that readers can understand what is supposed to
happen.
- `aggregate` is now only tested on a small model as it does not mean
anything to test it globally for all models.
- Previous tests are unchanged in desired output.
- Added a new test case that showcases better the difference between the
  FIRST, MAX and AVERAGE strategies.

Fixes #10263, #10763
See also #10568

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@LysandreJik

### CODE REVIEW

File: tests/test_pipelines_token_classification.py
Comment (francescorubbo) [reply tests/test_pipelines_token_classification.py:279]: Does this implementation support a scenario where we need to aggregate subwords into words, but not words into entities (e.g. in case the model labels do not follow the BIO convention we don't want to merge multiple entities with the same class)?
I think it doesn't, but I might have missed some details of the implementation logic.

**Comment by francescorubbo** [comment]
> @francescorubbo, @elk-cloner, this PR originated from yours and is based on the same approach. If you approve of this PR, can I add you as co-authors, as you've greatly contributed to its current shape?

Sure!

**Comment by LysandreJik** [comment]
Great, I will! 

Pinging also @joshdevins and @cceyda for feedback

File: src/transformers/pipelines/token_classification.py
Comment (cceyda) [inline src/transformers/pipelines/token_classification.py:310]: umm `average_scores.argmax()` would just return the index where max score occurs. which !=entity_idx on `self.model.config.id2label`
am I missing something? shouldn't it be:
```py
entity = entities[entity_idx]["entity"] # which is already self.model.config.id2label mapped
```

Comment (cceyda) [reply src/transformers/pipelines/token_classification.py:310]: also not sure what exactly is happening in the average strategy based on the code 😵 In the code `scores` looks like a list [0.1, 0.5 , 0.1, 0.4] which makes averaging then argmaxing make no sense.

in my mind avg is:
|I-ORG | I-PER | I-PER | I-LOC|
|---|---|---|---|
|0.1      |  0.5    |   0.1   |  0.4|

first strategy -> 0.1 I-ORG
max strategy -> 0.5 I-PER
**avg strategy -> (0.3 I-PER, 0.4 I-LOC 0.1 I-ORG )->0.4 I-LOC**
I guess there can also be a sum strategy -> (0.6 I-PER, 0.4 I-LOC 0.1 I-ORG ) -> 0.6 I-PER

(btw, scores in reality are floats so we wouldn't be concerned about tie breaking (in max) etc.)

**Review by cceyda — COMMENTED** [review]
~Other than the avg strategy,~
It all looks good! much cleaner than before too
Good job @elk-cloner @Narsil @francescorubbo 🥳 
wait a second I'm seeing something:
image
**Edit:** reached the right conclusion by the wrong means 🤣 
The above error is due to that subword getting predicted as a "O"(ignored default) and being ignored early on.
this bit of code [line](https://github.com/huggingface/transformers/blob/572f972092d8e38760558a0047d2e12cf79938bc/src/transformers/pipelines/token_classification.py#L245)
```py
 filtered_labels_idx = [
                (idx, label_idx)
                for idx, label_idx in enumerate(labels_idx)
                if (self.model.config.id2label[label_idx] not in self.ignore_labels) and not special_tokens_mask[idx]
            ]
```
we should probably do `ignore_labels` check later after we merge the subwords, or we will end up wrongfully ignoring some subwords' contributions in max|avg strategies.

File: src/transformers/pipelines/token_classification.py
Comment (francescorubbo) [reply src/transformers/pipelines/token_classification.py:310]: Isn't `scores` a 2D array of dimensions (N_subwords, N_softmax)? We are averaging along subwords and then taking the  overall argmax label. This makes sense if we assume each subword carries some amount of uncorrelated signal about the entity label.

Comment (cceyda) [reply src/transformers/pipelines/token_classification.py:310]: aha just realized, it all makes sense now!
Nevermind, morning brain  🤦

File: src/transformers/pipelines/token_classification.py
Comment (cceyda) [inline src/transformers/pipelines/token_classification.py:393]: just sayin `True` when `index` is not available feels dangerous 🧐 
non consecutive entities having matching B-I-tags it could happen.

Comment (cceyda) [reply src/transformers/pipelines/token_classification.py:393]: we might have to keep track of first subword index and last subword index of the entity during aggregation.
maybe index could be a tuple 🤔

Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:393]: You are perfectly correct, and doing the filtering later should actually remove that check anyway (as there can't be any more non-consecutive entities anymore.

File: tests/test_pipelines_token_classification.py
Comment (Narsil) [reply tests/test_pipelines_token_classification.py:279]: Ok, you're right that this currently does not support that, is that correct that in the absence of '-' within the Tag name we can assume everything is 'B-' ?

File: src/transformers/pipelines/token_classification.py
Comment (cceyda) [reply src/transformers/pipelines/token_classification.py:393]: oh that's right~ turns out it's a - kill two 🐛 with one 🥌- situation.

Comment (Narsil) [reply src/transformers/pipelines/token_classification.py]: It led to a bit more consequential change, but should be good now.

We are still filtering special_tokens early on, but `ignore_labels` later, so all contributions should be there.

File: tests/test_pipelines_token_classification.py
Comment (francescorubbo) [reply tests/test_pipelines_token_classification.py]: Maybe it's more robust to check explicitly in the label dict if they match the BIO pattern. Using a conditional on absence of '-' seems brittle as it could yield a false negative if a label is e.g. 'over-the-counter' (imagine a model classifying medications in prescription notes).

**Comment by joshdevins** [comment]
Looking good, thank you for all the work! I'm wondering if you can include test cases for the original 3 examples provided in https://github.com/huggingface/transformers/issues/10263#issue-811193366 ? The new test examples here look correct but I'm not sure they cover the scope of the first examples.

File: tests/test_pipelines_token_classification.py
Comment (LysandreJik) [inline tests/test_pipelines_token_classification.py:73]: We're in the process of renaming all `nlp` variables to more precise names (for example here `token_classifier`), if you want to take the opportunity of that refactor to rename those, that would be great!

File: tests/test_pipelines_token_classification.py
Comment (LysandreJik) [inline tests/test_pipelines_token_classification.py:197]: This example could be put on a single line to improve readability, for example like it's done here:

https://github.com/huggingface/transformers/blob/113eaa757589d31ee355c6c5d1fbd29e7e8ba788/tests/test_tokenization_tapas.py#L1188-L1190

File: tests/test_pipelines_token_classification.py
Comment (LysandreJik) [inline tests/test_pipelines_token_classification.py:434]: :+1:

**Review by LysandreJik — APPROVED** [review]
LGTM, thank you for this refactor! Let me know when satisfied and I'll operate the `merge` while adding the co-authors as mentioned above.

**Comment by Narsil** [comment]
@joshdevins 

Do you mind giving an example displaying the issue ?

I tried this, but I don't think it exhibits what you mention in the original issue:  ( https://github.com/huggingface/transformers/issues/10263#issue-811193366)

```python
        NER_MODEL = "elastic/distilbert-base-cased-finetuned-conll03-english"
        model = AutoModelForTokenClassification.from_pretrained(NER_MODEL)
        tokenizer = AutoTokenizer.from_pretrained(NER_MODEL, use_fast=True)
        sentence = """Accenture is a company. Max Mustermann is someone, Elasticsearch is something."""
        nlp_ner = pipeline("ner", model=model, tokenizer=tokenizer)
        output = nlp_ner(sentence)
        print(output)
        self.assertEqual(
            nested_simplify(output),
            [
                {"entity": "B-PER", "score": 0.9953969, "index": 9, "word": "Max", "start": 24, "end": 27},
                {"entity": "I-PER", "score": 0.9773876, "index": 10, "word": "Must", "start": 28, "end": 32},
                {"entity": "I-PER", "score": 0.9924896, "index": 11, "word": "##erman", "start": 32, "end": 37},
                {"entity": "I-PER", "score": 0.9860034, "index": 12, "word": "##n", "start": 37, "end": 38},
                {"entity": "B-ORG", "score": 0.99201995, "index": 16, "word": "El", "start": 51, "end": 53},
                {"entity": "B-ORG", "score": 0.99391395, "index": 17, "word": "##astic", "start": 53, "end": 58},
                {"entity": "B-ORG", "score": 0.9962443, "index": 18, "word": "##sea", "start": 58, "end": 61},
                {"entity": "B-ORG", "score": 0.9924281, "index": 19, "word": "##rch", "start": 61, "end": 64},
            ],
        )
```

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:160]: Done

File: tests/test_pipelines_token_classification.py
Comment (Narsil) [reply tests/test_pipelines_token_classification.py:197]: Perfect.

File: tests/test_pipelines_token_classification.py
Comment (Narsil) [reply tests/test_pipelines_token_classification.py:73]: Yes, I've done it !

File: tests/test_pipelines_token_classification.py
Comment (Narsil) [reply tests/test_pipelines_token_classification.py]: I've chnaged the code to be more restrictive.

**Comment by Narsil** [comment]
I think we should just wait for the test with `elastic`  model and we' re good to go.

**Comment by joshdevins** [comment]
@Narsil I've since retrained that model so it works better with new domain data like "Elasticsearch".

The point was more to decouple the fix from a specific model and to be robust to possible outputs of a model particularly for words/sub-words that are out-of-domain for a model and relying on sub-word token classification which can be less predictable (in my experience). This was why I suggested stubbing out the model and just putting in the sub-word labels directly to the aggregator to see if the expected behaviour matches the actual new behaviour.

**Comment by Narsil** [comment]
Yes, then I think I already added those tests here:

```python
    def test_aggregation_strategy_example2(self):
        model_name = self.small_models[0]
        tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
        nlp = pipeline(task="ner", model=model_name, tokenizer=tokenizer, framework="pt")
        # Just to understand scores indexes in this test
        self.assertEqual(
            nlp.model.config.id2label,
            {0: "O", 1: "B-MISC", 2: "I-MISC", 3: "B-PER", 4: "I-PER", 5: "B-ORG", 6: "I-ORG", 7: "B-LOC", 8: "I-LOC"},
        )
        example = [
            {
                # Necessary for AVERAGE
                "scores": np.array([0, 0.55, 0, 0.45, 0, 0, 0, 0, 0, 0]),
                "is_subword": False,
                "index": 1,
                "word": "Ra",
                "start": 0,
                "end": 2,
            },
            {
                "scores": np.array([0, 0, 0, 0.2, 0, 0, 0, 0.8, 0, 0]),
                "is_subword": True,
                "word": "##ma",
                "start": 2,
                "end": 4,
                "index": 2,
            },
            {
                # 4th score will have the higher average
                # 4th score is B-PER for this model
                # It's does not correspond to any of the subtokens.
                "scores": np.array([0, 0, 0, 0.4, 0, 0, 0.6, 0, 0, 0]),
                "is_subword": True,
                "word": "##zotti",
                "start": 11,
                "end": 13,
                "index": 3,
            },
        ]
        self.assertEqual(
            nlp.aggregate(example, AggregationStrategy.NONE),
            [
                {"end": 2, "entity": "B-MISC", "score": 0.55, "start": 0, "word": "Ra", "index": 1},
                {"end": 4, "entity": "B-LOC", "score": 0.8, "start": 2, "word": "##ma", "index": 2},
                {"end": 13, "entity": "I-ORG", "score": 0.6, "start": 11, "word": "##zotti", "index": 3},
            ],
        )

        self.assertEqual(
            nlp.aggregate(example, AggregationStrategy.FIRST),
            [{"entity_group": "MISC", "score": 0.55, "word": "Ramazotti", "start": 0, "end": 13}],
        )
        self.assertEqual(
            nlp.aggregate(example, AggregationStrategy.MAX),
            [{"entity_group": "LOC", "score": 0.8, "word": "Ramazotti", "start": 0, "end": 13}],
        )
        self.assertEqual(
            nested_simplify(nlp.aggregate(example, AggregationStrategy.AVERAGE)),
            [{"entity_group": "PER", "score": 0.35, "word": "Ramazotti", "start": 0, "end": 13}],
        )
```

**Comment by joshdevins** [comment]
@Narsil Ah cool, I missed those examples in my read-through. LGTM 🎉

**Comment by Narsil** [comment]
@LysandreJik I changed the co-authors, I'll merge after you check I've done it correctly.

**Comment by LysandreJik** [comment]
Looks good to me, feel free to merge!

**Pull request merged**
